test/CdefTest.cc

```diff
@@ -0,0 +1,558 @@
+                    pri_strength = 19;
+                /* second strength can only be 0, 1, 2, 4 for 8-bit */
+                for (sec_strength = 0; sec_strength <= 4 << (bd_ - 8);
+                     sec_strength += 1 << (bd_ - 8)) {
+                    if (sec_strength == 3 << (bd_ - 8))
+                        continue;
+                    cdef_ref_(bd_ == 8 ? (uint8_t *)dst_ref_ : 0,
+                              dst_ref_,
+                              size_,
+                              src_ + CDEF_HBORDER + CDEF_VBORDER * CDEF_BSTRIDE,
+                              pri_strength,
+                              sec_strength,
+                              dir,
+                              pri_damping,
+                              sec_damping,
+                              bsize_,
+                              0 /*Not used*/,
+                              bd_ - 8);
+                    cdef_tst_(bd_ == 8 ? (uint8_t *)dst_tst_ : 0,
+                              dst_tst_,
+                              size_,
+                              src_ + CDEF_HBORDER + CDEF_VBORDER * CDEF_BSTRIDE,
+                              pri_strength,
+                              sec_strength,
+                              dir,
+                              pri_damping,
+                              sec_damping,
+                              bsize_,
+                              0 /*not used*/,
+                              bd_ - 8);
+
+                    for (pos = 0; pos < max_pos; pos++) {
+                        ASSERT_EQ(dst_ref_[pos], dst_tst_[pos])
+                            << "Error: CDEFBlockTest, SIMD and C mismatch."
+                            << std::endl
+                            << "First error at " << pos % size_ << ","
+                            << pos / size_ << " (" << (int16_t)dst_ref_[pos]
+                            << " : " << (int16_t)dst_tst_[pos] << ") "
+                            << std::endl
+                            << "pristrength: " << pri_strength << std::endl
+                            << "pridamping: " << pri_damping << std::endl
+                            << "secstrength: " << sec_strength << std::endl
+                            << "secdamping: " << sec_damping << std::endl
+                            << "bitdepth: " << bd_ << std::endl
+                            << "size: " << bsize_ << std::endl
+                            << "boundary: " << boundary_ << std::endl
+                            << std::endl;
+                    }
+                }
+            }
+        }
+    }
+
+    void test_cdef(int iterations) {
+        int pri_damping, sec_damping, bits, level, count;
+        // for 8-bit content, damping ranges [3, 6] for luma, [2, 5] for chroma
+        // for highbd content, range will add the additinal bitdepth. 12-bit
+        // content will range [7, 10]
+        for (pri_damping = 3 + bd_ - 8;
+             pri_damping < 7 - 3 * !!boundary_ + bd_ - 8;
```

Comment:
can we move "7 - 3 * !!boundary_ + bd_ - 8" outside the loop and it seems constant during the processing in loop

Reply:
Done.